src/Symfony/Component/Messenger/Middleware/HandleMessageInNewTransactionMiddleware.php

```diff
@@ -0,0 +1,125 @@
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\QueuedMessageHandlingException;
+use Symfony\Component\Messenger\Stamp\Transaction;
+
+/**
+ * Allow to configure messages to be handled in a new transaction.
+ * I.e, messages dispatched from a handler with a Transaction stamp will actually be handled
+ * once the current message being dispatched is fully handled or sent.
+ *
+ * For instance, using this middleware before the DoctrineTransactionMiddleware
+ * means sub-dispatched messages with a Transaction item would be handled after
+ * the Doctrine transaction has been committed.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class HandleMessageInNewTransactionMiddleware implements MiddlewareInterface
+{
+    /**
+     * @var QueuedEnvelope[] A queue of messages and next middleware
+     */
+    private $queue = [];
+
+    /**
+     * @var bool Indicates if we are running the middleware or not. I.e, are we called during a dispatch?
+     */
+    private $isRunning = false;
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (null !== $envelope->last(Transaction::class)) {
+            if (!$this->isRunning) {
+                throw new \LogicException(sprintf('You can only use a "%s" stamp to define a new transaction in the context of a message handling.', Transaction::class));
+            }
+            $this->queue[] = new QueuedEnvelope($envelope, $stack);
+
+            return $envelope;
+        }
+
+        if ($this->isRunning) {
+            /*
+             * If come inside a second dispatch, just continue as normal.
+             * We should not run the stored messages until first call is finished.
+             */
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        // First time we get here, mark as inside a root dispatch call:
+        $this->isRunning = true;
+        try {
+            // Execute the whole middleware stack & message handling for main dispatch:
+            $returnedEnvelope = $stack->next()->handle($envelope, $stack);
+        } catch (\Throwable $exception) {
+            /*
+             * Whenever an exception occurs while handling a message that has
+             * queued other messages, we drop the queued ones.
+             * This is intentional since the queued commands were likely dependent
+             * on the preceding command.
+             */
+            $this->queue = [];
+            $this->isRunning = false;
+
+            throw $exception;
+        }
+
+        // Root dispatch call is finished, dispatch stored ones for real:
+        $exceptions = [];
+        while (null !== $queueItem = array_shift($this->queue)) {
+            try {
+                // Execute the stored messages
+                $queueItem->getStack()->next()->handle($queueItem->getEnvelope(), $queueItem->getStack());
+            } catch (\Throwable $exception) {
```

Comment:
I think we should differentiate `\Error` from `\Exception`: `Error should abort immediately and should be thrown unwrapped.

Reply:
Thanks